spec/controllers/admin/tags_controller_spec.rb

```diff
@@ -0,0 +1,21 @@
+require "rails_helper"
+
+describe Admin::TagsController, :admin do
+  describe "POST create" do
+    context "with tag name" do
+      it "redirects to /admin/tags" do
+        post :create, params: { tag: { name: "foo" } }
+  
+        expect(response).to redirect_to "/admin/tags"
+      end
+    end
+
+    context "without tag name" do
+      it "tag model returns error message" do
+        post :create, params: { tag: { name: "" } }
+        
+        expect(assigns(:tag).errors.messages[:name]).to eq ["can't be blank"]
+      end
+    end
+  end
+end
```

Comment:
We do not usually use controller tests to check user workflows. Instead, we use system tests which emulate users' behaviour. Probably we can replace the controller tests by adding a new system spec to `spec/system/admin/tags_spec.rb` that checks it's not possible to create a new tag when the name is blank.

Here is an example of a system spec that emulates the user workflow:

```
  scenario "Create shows validation error when tag name is empty" do
    visit admin_tags_path

    within("form.new_tag") do
      fill_in "tag_name", with: ""
      click_button "Create topic"
    end

    expect(page).to have_content "can't be blank"
  end
```

Reply:
I think that both tests are valid, cause `spec/system/admin/tags_spec.rb` I think is a e2e test and the `spec/controllers/admin/tags_controller_spec.rb` is an simple unit test for controller. What do you think?